### ISSUE TITLE
What is "wire"?

### ISSUE THREAD

**Issue by readevalprint** [tendermint/go-amino#3]
So I'm trying to make a Python3 compatible example for the tmsp. And came across this error[1] in the current Python 2 version. But reading through this repo I'm trying to understand what is the goal of making a new protocol. This seems really close to BSON but the benefit of BSON is that is supports hash mapping.

https://en.wikipedia.org/wiki/BSON

[1]

```
python ./tmsp/wire.py
Traceback (most recent call last):
  File "./tmsp/wire.py", line 94, in <module>
    ds = map(decode_big_endian, bs,ss)
  File "./tmsp/wire.py", line 33, in decode_big_endian
    firstByte = reader.read(1)[0]
AttributeError: 'bytearray' object has no attribute 'read'
```

**Comment by ebuchman** [comment]
Great question. This is an active area of discussion for us and maybe you can weigh in. We rolled our own serialization format, whose canonical implementation is at https://github.com/tendermint/go-wire. The most important thing is that the serialization be 100% deterministic - hash maps typically don't guarantee that since iteration is randomized, but maybe BSON does? Anyways, we've been thinking about probably just porting everything to protobuff soon. What do you think?

As for the error above, bytearrays are wrapped in the reader class to export a read method that shaves bytes off as they are read to facilitate stream encoding/decoding. If you have a superior approach, pull requests are very welcome!

**Comment by jaekwon** [comment]
To add to the discussion,

The goal is to be as short of a binary representation as reasonably possible while still being very simple to parse, while also keeping it deterministic.  BSON encodes the keys of fields, and in the case of arrays it encodes the array indices as well.  Wire only encodes the field values in order, and the array values in order as well.  Apache Avro or SBE are also comparables.

**Comment by readevalprint** [comment]
tl;dr protobuf is great if your data schema is known in advance.

Thanks for the response, I do understand the need for determinism. Coming from a Python background we have OrderedDict whose keys are sorted in the order added which allows you to then sort keys alphabetically. From a quick search, it seems golang sorts JSON keys lexicographically too.

As for Protobuf, I considered mentioning it but I wasn't sure of your needs. It's great because it provides validation but it's not great if you need arbitrarily structured data. Also check out https://capnproto.org/, it's from the author of Protobuf and he declares it the successor to it, smaller faster etc. 

Just a small suggestion, you all have a lot of work ahead of you so I really wouldn't suggest writing your own protocol for serialization because that's not this project's raison d'être. I'd even punt the decision on how to optimize the smallest representation until as late as possible.

Anyways this is great stuff, I'll keep plugging away and, if you don't mind, make a PR with the py3 version eventually.

**Comment by ebuchman** [comment]
Great info, thanks! We've been punting on what to do with wire but it has come up a few times now from users who are confused about it lol.  Will check out capnproto. All our data types are known ahead of time so I'm pretty sure protobuf or its successor would be a good fit for us. In the meantime, until its more urgent, we'll probably just stick with wire, and the necessary pieces for implementing in other languages are small enough that its pretty easy.

Looking forward to the py3 compatibility and any other improvements you might be able to make to the python server! Thanks!

**Comment by jtremback** [comment]
Protobufs is not deterministic, unfortunately
